### PR TITLE
Add SubPathExpr option for additionalVolumes

### LINKED ISSUE #2464: No option SubPathExpr for additionalVolumes
Please, answer some short questions which should help us to understand your problem / question better?

- **Which image of the operator are you using?**

registry.opensource.zalan.do/acid/postgres-operator:v1.10.0

- **Where do you run it - cloud or metal? Kubernetes or OpenShift?** [AWS K8s | GCP ... | Bare Metal K8s]

Kubernetes on private vms

- **Are you running Postgres Operator in production?** [yes | no]

no

- **Type of issue?** [Bug report, question, feature request, etc.]

Feature request

### PR DESCRIPTION
**Pull request by smutel**
Fix #2464

### CODE REVIEW

**Comment by smutel** [comment]
`SubPathExpr and SubPath are mutually exclusive` => I don't know how to implement this. Tell me if you have an idea.

**Comment by micabe** [comment]
Hey @hughcapet,
Can we merge this PR ? It's a useful option to have.
Thanx

**Comment by rds13** [comment]
👍 
Nice addition.

File: pkg/cluster/k8sres.go [hunk 1]
```diff
@@ -1752,9 +1752,10 @@ func (c *Cluster) addAdditionalVolumes(podSpec *v1.PodSpec,
 			for _, target := range additionalVolume.TargetContainers {
 				if podSpec.Containers[i].Name == target || target == "all" {
 					mounts = append(mounts, v1.VolumeMount{
-						Name:      additionalVolume.Name,
-						MountPath: additionalVolume.MountPath,
-						SubPath:   additionalVolume.SubPath,
+						Name:        additionalVolume.Name,
+						MountPath:   additionalVolume.MountPath,
+						SubPath:     additionalVolume.SubPath,
+						SubPathExpr: additionalVolume.SubPathExpr,
```
Comment (FxKu) [inline pkg/cluster/k8sres.go:1758]: If SubPath and SubPathExpr are mutually exclusive they cannot coexist here, right?

**Comment by FxKu** [comment]
This PR is still lacking a few things. The CDR needs to be extended [here](https://github.com/zalando/postgres-operator/blob/master/manifests/postgresql.crd.yaml) and [here](https://github.com/zalando/postgres-operator/blob/master/pkg/apis/acid.zalan.do/v1/crds.go). If the SubPath and SubPathExpr are mutually exclusive you can use `OneOf` syntax, I think. So far, we are not using it anywhere so you have to search for examples online.

When adding fields to the Postgres CRD, please also run code generation (`./hack/update-codegen.sh`). And mind that we also have the helm chart, where this new feature has to be addressed, too.

**Comment by smutel** [comment]
Hello @FxKu,

Thank you for you review.
I did the changes you ask me (I hope).

**Comment by smutel** [comment]
I am facing this https://github.com/kubernetes/apimachinery/issues/18.

**Comment by smutel** [comment]
Sounds good now.

**Comment by smutel** [comment]
Any news on this ?

File: pkg/apis/acid.zalan.do/v1/crds.go [hunk 2]
```diff
@@ -167,6 +167,9 @@ var PostgresCRDResourceValidation = apiextv1.CustomResourceValidation{
 									"subPath": {
 										Type: "string",
 									},
+									"subPathExpr": {
+										Type: "string",
+									},
```
Comment (FxKu) [inline pkg/apis/acid.zalan.do/v1/crds.go:172]: can you reflect the oneOf part also in the Go code?

Comment (smutel) [reply pkg/apis/acid.zalan.do/v1/crds.go:182]: Done.

**Comment by FxKu** [comment]
@smutel e2e tests are failing. There's an issue with the CRD schema:
```
spec.validation.openAPIV3Schema.properties[spec].properties[additionalVolumes].items.oneOf[0].properties[subPath].type: Forbidden: must be empty to be structural, 
spec.validation.openAPIV3Schema.properties[spec].properties[additionalVolumes].items.oneOf[1].properties[subPathExpr].type: Forbidden: must be empty to be structural
```

**Comment by smutel** [comment]
CRDs updated.

**Comment by FxKu** [comment]
Btw. I have just noticed that you did not add any logic to the Go code. It's not enough to just add a new field to the CRD. We use a lot of custom logic and not so much K8s specs. In this case the new SubPathExpr muss be set in [generateVolumeMounts](https://github.com/zalando/postgres-operator/blob/master/pkg/cluster/k8sres.go#L638) function. I would like to see a unit test then to check if the new expression is set when specified.

Sorry for suggesting oneOf here as both fields are optional. It's more typical when you have a required `type` field which specifies different options. You could even put the SubPathExpr in the SubPath field and use an extra bool flag `isSubPathExpr`. That makes implementation more straight forward. Hope you understand what I mean here.

**Comment by smutel** [comment]
I changed my PR regarding your suggestion with boolean.
Perhaps the default value for this boolean is missing but I don't know where to specify this.

File: pkg/cluster/k8sres_test.go [hunk 3]
```diff
@@ -1865,7 +1884,7 @@ func TestAdditionalVolume(t *testing.T) {
 		{
 			subTest:        "checking volume mounts of postgres container",
 			container:      constants.PostgresContainerName,
-			expectedMounts: []string{"pgdata", "test1", "test3", "test4"},
+			expectedMounts: []string{"pgdata", "test1", "test3", "test4", "test5", "test6"},
```
Comment (FxKu) [inline pkg/cluster/k8sres_test.go:1887]: Please extend this unit test to by adding `expectedSubPath` and `expectedSubPathExpr` to the test struct so you can compare it later when looping over the test cases.

**Comment by smutel** [comment]
Looks good to you ?

**Comment by smutel** [comment]
Any update on this ?

**Comment by FxKu** [comment]
LGTM now. Sorry for not getting back to this sooner.

**Comment by FxKu** [comment]
:+1:

**Pull request merged**